### ISSUE TITLE
Port to Moose 11

### ISSUE THREAD

**Issue by fuhrmanator** [fuhrmanator/FamixTypeScript#16]
Pharo 12 will soon be finalized, and so Moose 11 should be stable soon as well. This means the FamixTypeScript MM should work in Moose 11.

Currently, here are some of the problems:

- [x] TParameterType in Moose 11 doesn't exist, and so the MM can't be generated, for example (in `defineHierarchy`):
    ```st
	    parameterType --|> type.
	    parameterType --|> #TParameterType.
    ```
    There are surely hints in the FamixJava metamodel in Moose 11 ;-) So far, I've found:
    - there are classes (without traits) `ParametricClass`, `ParameterType`, `ParametricInterface`, `ParametricMethod`, and `ParameterConcretization` (perhaps unrelated but seems like it's important). 

- [x] At least one class in Extensions has a dependency on a class that's generated in Entities. When I try to load Extensions, I get the following Warning (Maybe this can be fixed in the Baseline? Maybe this is just a stricter warning in Pharo 11?):
  > Warning: Package *Famix-TypeScript-Extensions depends on the following classes:
  >  FamixTypeScriptParameterizableClass
  >  FamixTypeScriptParameterType
  >  FamixTypeScriptType
  >  FamixTypeScriptClass
  > You must resolve these dependencies before you will be able to load these definitions: 
  >  FamixTypeScriptClass>>#accessedAttributes
  >  FamixTypeScriptClass>>#isInterface
  >  FamixTypeScriptClass>>#isInterface:
  >  FamixTypeScriptClass>>#methodsWithoutSutbsAndConstructors
  >  FamixTypeScriptParameterType>>#isParameterType
  >  FamixTypeScriptParameterizableClass>>#parameters
  >  FamixTypeScriptParameterizableClass>>#parameters:
  >  FamixTypeScriptType>>#isInterface
  >  FamixTypeScriptType>>#isParameterType

**Comment by fuhrmanator** [comment]
@nicolasanquetil Feel free to add items you think will help here :)

**Comment by fuhrmanator** [comment]
To gain insights into the changes in Famix (Traits, etc.) for Moose 11, [here's a PlantUML class diagram of the Moose 11 version of FamixJava metamodel](https://www.plantuml.com/plantuml/svg/fLZDRjiu4BxhAOJiPNUnwMcFHT74Q1OfsZJOkEcP9YVsmDJG659Ed7RVVKEnCaLJYWYqbqGp7-V_PvhtrWbsRQEh3IgeeTcvGzr0imAsbTqYxGIBffPQM5j_h-hwKiZjcarBQc6quVfnWmvUw1VCvZ6GLle5uw665N2WAS7RVrWSgl-g8EkwsMcKwEu5erXfgD--_NtnLoVjZtVlwj-kkd_rz-43msuMjbpu7v_1CKhhqSinT5tVjDx7cheVNnvksvL6-V9iOvGTORDn81seLn2u5muc-Gj3rd4hdU59p8MKO6q4z3xT2haLw-aGvSeq0hcVpqOKPHOmQE5m3o78WYHS6RuxM0TD1C730p2GMvXsfv7MaIDMrh6GZi0EX3Q27TB2RC2YeKXNQ3iLW_UJabSjoKqk0iawiKADBbRKLzgIUQGxqxA4JuBMxUinktECjAv9D72KSK5ad71ew9g2aRC5rzNRBEhMjn4uuELXDo_pmZd6LUlcaQ-AyoreD0YgqLuJ0NUs9kGbMBSG5XBYnw-4xk-U5JkG53o1ka8D5oGtXcUj-vLSn90j34a6XyzTOcQLVZ9IwBAn4hKmoh94BY_zx8k3PGBs3RMIWjMil0_KDhEWM9O-tuNL4LiY1gCeN-e3ENI7yd7pKUp5JQlrOWDoQmVriW76iYuA4wceahiyNHCw51gVWQVcAAILxXSED40EQPqmlko16LKWPrExyK_6vxb2dn6XxKJpZ6bHGBciQzgRRTmjurcVhsJ61lqGZ-YYkWzgPnzu8sLP1ov30IqOX3CysojNei4dNqEVZG9TO8W3VX0IYdhhNkWMoiVpBF9Vw5PLWVRbGj0bt1k_MKFlBUTAkmUVDEGIdboSk8LfBGzJQ9T2VvOio2AGAtBeY8uvUPAmcwsYw7kXtwKx8l7bxa1omuRm6LIHEyC7GzLE8067gf_N_GtTvfWT6nJFRXnGiy2VmshMWFp8ATyenobQ5kSRqO0gZCiiB17su_my-pz4TS8qxAgOdSZh3-eglKy2BpyZ8pBHdeUn2fFNCxib5R_J-o_tv9Y0SNkxHqa_U79-a2K1I3t8TIPtKl9oulAgEgSI2ACyzTGFtVXvF8Ho_HD7JGuykMAYy7JbvSxtOX9El4ecWXtsOTHoOcWL-ATX2DH4KzoOKLZSNfcVSLqDfSI5anT0jXuocJ6LrL5nPAIB8bKRbi8WvN5ovu4UAx-qTdBBJc8_qc7zG2jhlbX0GrmwhyOa3GohJlvuyEFDcxwHqzqNM55ZkkepRZV00xMhe420G8eTaY-b0EWBChcg0YizdQfijNZCIQ1xSywqUGJkVZ-hpy-7tw1wc-oVPyVzqnU-n-BmOlMWyDjHThlR9RBZzuJiYM7GFTu91d8fuYplojGuf7LctG0ZjS_Z9Mp47atBuw0Tcpqgu37OUo3LDhhw7m00), generated with the following script in a Playground:

```st
FamixUMLDocumentor new
  model: FamixJavaModel ;
  generate ;
  exportWith: (FamixUMLPlantUMLBackend new).
```

**Comment by fuhrmanator** [comment]
Here are some examples with Generics:

###  Generic Function
A generic function can accept parameters of any type and return a value of the same type:

```typescript
function identity<T>(arg: T): T {
    return arg;
}

let numberIdentity = identity<number>(42); // numberIdentity is of type number
let stringIdentity = identity<string>("Hello, TypeScript!"); // stringIdentity is of type string
let arrayIdentity = identity<number[]>([1, 2, 3]); // arrayIdentity is of type number[]
```

### Generic Interface

A generic interface can define a structure that works with any type:

```typescript
interface GenericIdentityFn<T> {
    (arg: T): T;
}

function identityFn<T>(arg: T): T {
    return arg;
}

let myIdentity: GenericIdentityFn<number> = identityFn;
console.log(myIdentity(10)); // Outputs: 10
```

### Generic Class

A generic class can create instances that operate on various types:

```typescript
class GenericNumber<T> {
    zeroValue: T;
    add: (x: T, y: T) => T;

    constructor(zeroValue: T, add: (x: T, y: T) => T) {
        this.zeroValue = zeroValue;
        this.add = add;
    }
}

let myGenericNumber = new GenericNumber<number>(0, (x, y) => x + y);
console.log(myGenericNumber.add(5, 10)); // Outputs: 15

let stringNumeric = new GenericNumber<string>("", (x, y) => x + y);
console.log(stringNumeric.add("Hello, ", "world!")); // Outputs: Hello, world!
```

### Generic Constraints
You can add constraints to generics to restrict the kinds of types you can use:

```typescript
interface Lengthwise {
    length: number;
}

function loggingIdentity<T extends Lengthwise>(arg: T): T {
    console.log(arg.length); // Now we know it has a .length property, so no error
    return arg;
}

loggingIdentity({ length: 10, value: "Hello" }); // Works because the argument has a length property
// loggingIdentity(3); // Error: Argument of type 'number' is not assignable to parameter of type 'Lengthwise'.
```

### Generic Types in Arrow Functions
You can also use generics with arrow functions:

```typescript
const identity = <T>(arg: T): T => {
    return arg;
};

let output = identity<string>("Hello, TypeScript!"); // output is of type string
console.log(output); // Outputs: Hello, TypeScript!
```

### Using Multiple Type Variables
You can define a function or class with multiple type variables:

```typescript
function map<K, V>(array: K[], fn: (item: K) => V): V[] {
    return array.map(fn);
}

let stringToNumber = map<string, number>(["1", "2", "3"], (str) => parseInt(str));
console.log(stringToNumber); // Outputs: [1, 2, 3]
```

**Comment by fuhrmanator** [comment]
Currently, ArrowFunction is not in the metamodel. #22

**Comment by fuhrmanator** [comment]
See https://github.com/moosetechnology/Famix/issues/773#event-12896692862 as well.

**Comment by fuhrmanator** [comment]
https://modularmoose.org/moose-wiki/Developers/predefinedEntities#genericity helps with understanding Generics in Java.

**Comment by fuhrmanator** [comment]
fixed in #18
